src/main/java/com/chrosciu/bootcamp/tasks/github/GithubClient.java

```diff
@@ -4,28 +4,38 @@
 import com.chrosciu.bootcamp.tasks.github.dto.Repository;
 import lombok.RequiredArgsConstructor;
 import reactor.core.publisher.Flux;
+import reactor.core.scheduler.Scheduler;
+import reactor.core.scheduler.Schedulers;
 
 @RequiredArgsConstructor
 public class GithubClient {
     private final GithubApi githubApi;
 
     public Flux<Repository> getUserRepositories(String username) {
-        //TODO: Implement
-        return null;
+        return githubApi.getUserRepositories(username)
+                .flatMapMany(Flux::fromIterable);
     }
 
     public Flux<Branch> getUserRepositoryBranches(String username, String repo) {
-        //TODO: Implement
-        return null;
+        return githubApi.getUserRepositoryBranches(username, repo)
+                .flatMapMany(Flux::fromIterable);
     }
 
     public Flux<Repository> getUsersRepositories(Flux<String> usernames) {
-        //TODO: Implement
-        return null;
+        return usernames.flatMap(this::getUserRepositories);
     }
 
     public Flux<String> getAllUserBranchesNames(String username) {
-        //TODO: Implement
-        return null;
+        return getUserRepositories(username)
+                .flatMap(repo -> getUserRepositoryBranches(username, repo.getName()))
+                .map(Branch::getName);
+    }
+
+    public Flux<String> getAllUserBranchesNamesParallel(String username) {
+        Scheduler scheduler = Schedulers.boundedElastic();
+        return getUserRepositories(username)
+                .subscribeOn(scheduler)
```

Comment:
To nie zrównolegli operacji :) - zmiana schedulera przerzuci emitowanie elementów do innego wątka - ale dalej będą one emitowane jeden po drugim (wynika to ze specyfikacji Reactive Streams, której niestety nie udało się omówić we wszystkich szczegółach).

Reply:
Racja, to będzie tylko "nieblokujące". Wymyśliłem coś takiego, ale pewniei tak się okaże, że należy to robić inaczej :D

    public Flux<String> getAllUserBranchesNamesParallel(String username) {
        return getUserRepositories(username)
                .parallel(4)
                .runOn(Schedulers.parallel())
                .flatMap(repo -> getUserRepositoryBranches(username, repo.getName()))
                .map(Branch::getName)
                .sequential();
    }

Reply:
Ha - duży plus za to że chciało Ci się tego poszukać :) i jest to krok w dobrą stronę, natomiast ParallelFlux jest tu trochę strzałem z armaty do komara :) Są tu bowiem nieco słabsze strony rozwiązania:
- użycie tylko 4 równoległych szyn, a zdecydowanie można by więcej (strzały po HTTP są robione w sposób nieblokujący, więc nie ma się tu co ograniczać), tylko wtedy znowu problemem jest scheduler - patrz niżej
- scheduler typu parallel mający sztywną pulę wątków co nie pozwoli na zbyt duże zrównoleglenie operacji, lepszy byłby tutaj elastic
- sequential na końcu, który powoduje że i tak elementy będą szeregowane podczas publikowania - czyli w pewnym sensie trochę zaprzecza paralelizacji
Jakby to wszystko zsumować to można by to zrobić trochę lepiej:
```
return getUserRepositories(username)
                .parallel(128)
                .runOn(Schedulers.elastic())
                .flatMap(repo -> getUserRepositoryBranches(username, repo.getName()))
                .map(Branch::getName)
                .sequential();
```
Ale z kolei takie podejście w praktyce jest równoważne temu by zrzucić to zrównoleglenie na samego flatMap-a - i wtedy można to po prostu zapisać tak:
 ```
   return getUserRepositories(username)
                .flatMap(repo -> getUserRepositoryBranches(username, repo.getName()).subscribeOn(Schedulers.elastic()))
                .map(Branch::getName);
```
Prawda że prościej ? :)